compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirLocalVariableAssignmentAnalyzer.kt

```diff
@@ -0,0 +1,349 @@
+                concurrentlyAssignedLocalVariablesStack.last() += it
+            }
+            else -> {
+                // no additional stuff to do for other cases
+            }
+        }
+    }
+
+    fun enterFunctionCallWithMultipleLambdaArgs(lambdaArgs: List<FirAnonymousFunction>) {
+        concurrentLambdaArgsStack.add(lambdaArgs.toSet())
+    }
+
+    fun exitFunctionCallWithMultipleLambdaArgs() {
+        concurrentLambdaArgsStack.removeLast()
+    }
+
+    companion object {
+        /**
+         * Computes assigned local variables in each execution path. This analyzer runs before BODY_RESOLVE. Hence, it works on
+         * syntactical information only.
+         *
+         * # Note on implementation detail
+         *
+         * The analyzer constructs a mini control flow graph that captures forking of execution path. Any conditional branches, declaration of
+         * lambda and local functions are forks. The mini CFG does not care about loop structures and effectively treats it as a linear sequence of
+         * statements. This is sufficient for the purpose of collecting unstable local variables. Similarly for try/catch/finally constructs.
+         *
+         * Also, for simplicity, all conditionals are treated as non-exhaustive. Hence, a fallback edge is always added along a conditional
+         * structure.
+         *
+         * While building the mini CFG, inside each node, we collect local variables that are assigned later in the execution path.
+         *
+         * For example, consider the following code.
+         *
+         * ```
+         * fun test() {
+         *   var x: Int = 0
+         *   if (true) {
+         *     run {
+         *       x = 1
+         *       var y = 1
+         *       y = 2
+         *     }
+         *   } else {
+         *     x = 2
+         *   }
+         * }
+         * ```
+         *
+         * The generated mini CFG looks like
+         *
+         * ┌───────┐
+         * │  if   │
+         * └─┬─┬─┬─┘
+         *   │ │ │ fallback
+         *   │ │ └───────────────────────────┐
+         *   │ │ false                       │
+         *   │ └───────────────────────┐     │
+         *   │ true                    │     │
+         * ┌─┴─────┐               ┌───┴───┐ │
+         * │  run  │               │ empty │ │
+         * └─┬───┬─┘               └───┬───┘ │
+         *   │   │ normal execution    │     │
+         *   │   └─────────────┐       │     │
+         *   │ normal lambda   │       │     │
+         * ┌─┴──────┐      ┌───┴───┐   │     │
+         * │ lambda │      │ empty │   │     │
+         * │  body  │      │       │   │     │
+         * └────────┘      └───┬───┘   │     │
+         *   ┌─────────────────┘       │     │
+         *   │ ┌───────────────────────┘     │
+         *   │ │ ┌───────────────────────────┘
+         * ┌─┴─┴─┴─┐
+         * │ after │
+         * │  if   │
+         * └───────┘
+         *
```

Comment:
Thanks for the nice ascii drawing!

IIUC, such node corresponds to `MiniFlow`. Could you also add notable changes in internal data per flow? e.g., what are recorded at `assignedLocalVariables` at each flow.

Reply:
Sounds good! Let me add more information to it.